File: cs/Op.cs
namespace Morth;

public class Op
{
    public OpCode Code { get; }
    public ulong Value { get; } = 0;

    private Op(OpCode code)
    {
        Code = code;
    }

    private Op(OpCode code, ulong value)
    {
        Code = code;
        Value = value;
    }

    public static Op Push(ulong value)
    {
        return new Op(OpCode.Push, value);
    }

    public static Op Plus()
    {
        return new Op(OpCode.Plus);
    }

    public static Op Minus()
    {
        return new Op(OpCode.Minus);
    }

    public static Op Dump()
    {
        return new Op(OpCode.Dump);
    }
}

File: zig/src/op.zig
const std = @import("std");
const Token = @import("token.zig").Token;

const Allocator = std.mem.Allocator;

pub const Op = struct {
    token: Token,
    code: OpCode,

    pub const OpCode = union(enum) {
        Push: u64,
        Plus,
        Minus,
        Equal,
        Dump,
        If: ?usize,
        Else: ?usize,
        End: ?usize,
        Dup,
        Gt,
        While,
        Do: ?usize,
        Mem,
        Load,
        Store,
        Syscall1,
        Syscall2,
        Syscall3,
        Syscall4,
        Syscall5,
        Syscall6,
        Dup2,
        Lt,
        Swap,
        Drop,
        Shr,
        Shl,
        Bor,
        Band,
    };

    const Self = @This();
    pub fn init(token: Token, code: OpCode) Self {
        return Self{ .token = token, .code = code };
    }

    pub fn deinit(self: *Self, allocator: Allocator) void {
        self.token.deinit(allocator);
    }
};
